[Support] Move raw_ostream::tie to raw_fd_ostream

aengelke opened the pull request:
Originally, tie was introduced by D81156 to flush stdout before writing to stderr. 030897523 reverted this due to race conditions. Nonetheless, it does cost performance, causing an extra check in the "cold" path, which is actually the hot path for raw_svector_ostream. Given that this feature is only used for errs(), move it to raw_fd_ostream so that it no longer affects performance of other stream classes.

aengelke committed:
[Support] Remove raw_ostream::tie

Originally, tie was introduced by D81156 to flush stdout before writing
to stderr. 030897523 reverted this due to race conditions. Nonetheless,
it does cost performance, causing an extra check in the "cold" path,
which is actually the hot path for raw_svector_ostream. Given that this
feature sees almost no use, remove it.

Reverts commit 1ce831912c797df1cb6d313d8e576a3f86175b6d.

jh7370 commented:
> What was the motivation to keep this in llvm-dwarfdump?

If warnings or errors are printed by llvm-dwarfdump without this, we end up with corrupted output, where stderr and stdout are intermixed in a semi-arbitrary way. It was a while back that I made the original change mind you, so I don't have the concrete example in front of me any more. Looking at discussions in the original reviews on related commits make me think llvm-dwarfdump's --debug-line option was particularly badly affected: I'd be surprised if there aren't tests that start failing (potentially flakily, given that it'll depend on buffer sizes and therefore the amount of data written, which may be platform-specific etc) because of this change.

In practice, I wanted this enabled for all tools however, it just never happened.

>  Is there any other way to achieve the same goal?

Unless the additional check is causing serious issues, I'm not sure this is the right question (and even if it is the right question, I think it's on you as the person proposing to remove the change to provide an alternative - I don't really work on LLVM beyond reviewing these days, so I don't have the time to consider alternatives to work that is currently doing the job well for at least one initial use-case I care about). What I feel the right initial question is, is this check actually making a significant difference to performance in real-world use cases?

aengelke committed:
Retain tie(), move to raw_fd_ostream, expand comment

aengelke commented:
> What I feel the right initial question is, is this check actually making a significant difference to performance in real-world use cases?

raw_ostream::write shows up in my (downstream application) profiles at ~0.6%, mostly through raw_svector_ostream, which is always unbuffered. The performance improvement of this one-condition-removal is [measureable (c-t-t)](http://llvm-compile-time-tracker.com/compare.php?from=8a25bb9b391bd13d824b1df43187b4c304011cee&to=04b794a8a02f06803f96c40de042b13fd4b12c5f&stat=instructions:u).

But I see your point. I reworked this so that tie() still works, but only on raw_fd_ostream (errs() is an raw_fd_ostream), so that other streams (svector, string) get performance gains. I also updated the comment to state the motivation/use case for this feature.

MaskRay commented:
Unneeded `tie` for non-fd streams do cause performance issues. I recall that `flush_tied_then_write` contributed noticeable overhead in MC: https://reviews.llvm.org/D145791#4185609
